src/beamlime/applications/daemons.py

```diff
@@ -0,0 +1,173 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+import asyncio
+from contextlib import contextmanager
+from typing import Any, Awaitable, Callable, Coroutine, Generator, List, Optional
+
+from ._parameters import ChunkSize, DataFeedingSpeed
+from ._random_data_providers import RandomEvents
+from ._workflow import Events
+from .base import BeamlimeMessage, DaemonInterface
+from .handlers import RawDataSent
+
+
+class MessageRouter(DaemonInterface):
+    """A message router that routes messages to handlers."""
+
+    message_pipe: List[BeamlimeMessage]
+
+    class StopRouting(BeamlimeMessage):
+        ...
+
+    def __init__(self):
+        self.handlers: dict[type, List[Callable[[BeamlimeMessage], Any]]] = dict()
+        self.awaitable_handlers: dict[
+            type, List[Callable[[BeamlimeMessage], Awaitable[Any]]]
+        ] = dict()
+        self.message_pipe = list()
+        self._break_routing_loop = False  # Break the routing loop flag
+
+    def break_routing_loop(self, message: Optional[BeamlimeMessage] = None) -> None:
+        if not isinstance(message, self.StopRouting):
+            raise TypeError(
+                f"Expected message of type {self.StopRouting}, got {type(message)}."
+            )
+        self._break_routing_loop = True
+
+    @contextmanager
+    def _handler_wrapper(
+        self, handler: Callable[..., Any], message: BeamlimeMessage
+    ) -> Generator[None, None, None]:
+        import warnings
+
+        try:
+            self.debug(f"Routing event {type(message)} to handler {handler}...")
+            yield
+        except Exception as e:
```

Comment:
This ultimately ignores the error. So it could end up running the handler at every update cycle even though it fails every time. Intentional?

Reply:
It was intentional but I think it's not a good idea... changing...